plonky2/src/gadgets/ecdsa.rs

```diff
@@ -0,0 +1,132 @@
+use std::marker::PhantomData;
+
+use crate::curve::curve_types::Curve;
+use crate::field::extension_field::Extendable;
+use crate::gadgets::arithmetic_u32::U32Target;
+use crate::gadgets::biguint::BigUintTarget;
+use crate::gadgets::curve::AffinePointTarget;
+use crate::gadgets::nonnative::NonNativeTarget;
+use crate::hash::hash_types::RichField;
+use crate::hash::poseidon::PoseidonHash;
+use crate::iop::target::{BoolTarget, Target};
+use crate::plonk::circuit_builder::CircuitBuilder;
+
+pub struct ECDSASecretKeyTarget<C: Curve>(NonNativeTarget<C::ScalarField>);
+pub struct ECDSAPublicKeyTarget<C: Curve>(AffinePointTarget<C>);
+
+pub struct ECDSASignatureTarget<C: Curve> {
+    pub r: NonNativeTarget<C::ScalarField>,
+    pub s: NonNativeTarget<C::ScalarField>,
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> CircuitBuilder<F, D> {
+    pub fn hash_to_bits(&mut self, x: Target, num_bits: usize) -> Vec<BoolTarget> {
+        let inputs = vec![x];
+        let hashed = self.hash_n_to_m::<PoseidonHash>(inputs, 1, true)[0];
+        self.split_le(hashed, num_bits)
+    }
+
+    pub fn hash_to_scalar<C: Curve>(
+        &mut self,
+        x: Target,
+        num_bits: usize,
+    ) -> NonNativeTarget<C::ScalarField> {
+        let h_bits = self.hash_to_bits(x, num_bits);
+
+        let two = self.two();
+        let mut rev_bits = h_bits.iter().rev();
+        let mut sum = rev_bits.next().unwrap().target;
+        for &bit in rev_bits {
+            sum = self.mul_add(two, sum, bit.target);
+        }
+        let limbs = vec![U32Target(sum)];
+        let value = BigUintTarget { limbs };
+
+        NonNativeTarget {
+            value,
+            _phantom: PhantomData,
+        }
+    }
+
+    pub fn verify_message<C: Curve>(
+        &mut self,
+        msg: Target,
+        sig: ECDSASignatureTarget<C>,
+        pk: ECDSAPublicKeyTarget<C>,
```

Comment:
I think we also need to check that `pk.0` is on the curve, with `self.curve_assert_valid(pk.0)` for example.